✨Add templating interface to clusterctl

wfernandes opened the pull request:
<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, minor or feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🏃 (:running:, other) -->

**What this PR does / why we need it**:
- This PR creates an interface (`YamlProcessor`) to allow for future extensible templating processors. 
- Currently, there is no change in behavior which `clusterctl` commands or usage. 
- The existing templating behavior of variable substitution is encapsulated in the `SimpleProcessor` which is also the default if no yaml processor are defined.
- This PR only exposes `YAMLProcessor` as part of the `clusterctl.GetClusterTemplateOptions`, that is, the yaml processor is exposed via the clusterctl client library for `clusterctl.GetClusterTemplate(opts)`.
- There are other minor changes and refactors.

**Note**
I'm happy to break this PR into smaller PRs. As of now, I kinda broke into separate commits.

**Release Notes:**
TODO

/area clusterctl

vincepri commented:
/assign

ncdc commented:
@fabriziopandini would you have time to review?

fabriziopandini commented on cmd/clusterctl/client/repository/components.go:
```diff
@@ -194,18 +189,23 @@ type ComponentsOptions struct {
+	}
+
+	processedYaml := rawyaml
+	if !options.SkipVariables {
```
I'm not sure about this test.
SkipVariables imply to process the template, but to not fail in case some variable is missing; instead, if I got this right this is changing the semantic of the flag by skipping template processing in this case

wfernandes replied:
It basically states that if you don't need variable substitution just return the original components yaml. But if variable substitution is required then process the yaml with variables.
For the existing Simple processor the processing is variable substitution that's why it seems like the `SkipVariables` option is used to gate the processing.
In the future if there are other types of processing for the infrastructure components we can refactor this then.

fabriziopandini replied:
Ok, I don't want to block on this and so l think that we can drop the partial variable substitution. 
Hopefully, this won't hurt users, and we land in a cleaner state.

fabriziopandini commented on cmd/clusterctl/client/repository/template.go:
```diff
@@ -70,23 +71,33 @@ func (t *template) Yaml() ([]byte, error) {
-	variables := inspectVariables(rawYaml)
+// TODO: Refactor to reduce the number of args passed in.
+func NewTemplate(
+	rawArtifact []byte,
```
What about introducing an input struct also here?

vincepri commented:
/hold

Holding to review/merge after we cut alpha.0 next week

wfernandes commented:
/retest

wfernandes commented on cmd/clusterctl/client/repository/components_client_test.go:
```diff
@@ -277,3 +312,56 @@ func Test_componentsClient_Get(t *testing.T) {
 	}
 }
+
+func Test_componentsClient_YamlProcessor(t *testing.T) {
```
I converted them into a table driven test. Hopefully that makes them more clear. Good catch for a clean refactor.

wfernandes commented:
/test pull-cluster-api-e2e

wfernandes commented:
@vincepri Can we remove the hold on this PR now that alpha is cut?

vincepri commented:
I have some time today to review

wfernandes commented on cmd/clusterctl/client/repository/components.go:
```diff
@@ -194,18 +189,23 @@ type ComponentsOptions struct {
-func NewComponents(provider config.Provider, configClient config.Client, rawyaml []byte, options ComponentsOptions) (*components, error) {
-	// Inspect the yaml read from the repository for variables.
-	variables := inspectVariables(rawyaml)
+func NewComponents(provider config.Provider, configClient config.Client, processor yaml.Processor, rawyaml []byte, options ComponentsOptions) (*components, error) {
```
Hmm...I thought I did it after your initial review but I guess I didn't. I'll add an Input struct in.

wfernandes replied:
Oh...yeah I added an input struct `repository.TemplateInput` for `NewTemplate` in `repository/template.go`.

fabriziopandini replied:
This comment is for NewComponents...

wfernandes commented on cmd/clusterctl/client/repository/template_client_test.go:
```diff
@@ -184,3 +218,40 @@ func Test_templates_Get(t *testing.T) {
 	}
 }
+
+type FakeProcessor struct {
```
Hahaha...yeah I tried really hard doing this  many times. 😆 

I would run into cyclic import issues because the `YamlProcessor` interface has `Process([]byte, config.VariablesClient) ([]byte, error)`. And the `config` package imports `internal/test` (since we don't use separate `_test` packages).

Even if I created my own interface for the variables client, it would have to be in a package that doesn't import `internal/test` which seemed wrong.

I was very sad 😞 to have to re-implement these fake processors in the different places.

fabriziopandini replied:
Looks like having a dependency on config for re-using an interface that exposes a single func is a bad idea...
What about having a YamlProcessor func interface, and somehere making an asserting the two interfaces actually met?

wfernandes replied:
> Looks like having a dependency on config for re-using an interface that exposes a single func is a bad idea...

I wish I thought of this sooner. 😄 

Yeah...so I was thinking about something like 
`Process([]byte, func(string) string) ([]byte, error)`. That's the most generic form of a "Getter" I could think of.

This way we could implement whatever getter we want.
```
varsGetter := func(key string) string {
  value, _ := configVariablesClient.Get(key)
  return value
}
processedYaml, err := Process(rawArtifact, varsGetter)
```

WDYT about this style?

vincepri commented on cmd/clusterctl/client/client_test.go:
```diff
@@ -118,23 +119,23 @@ func newFakeClient(configClient config.Client) *fakeClient {
+		k := cluster.Kubeconfig(i.kubeconfig)
 		if _, ok := fake.clusters[k]; !ok {
-			return nil, errors.Errorf("Cluster for kubeconfig %q and/or context %q does not exists.", i.Path, i.Context)
+			return nil, errors.Errorf("Cluster for kubeconfig %q and/or context %q does not exists.", i.kubeconfig.Path, i.kubeconfig.Context)
```
```suggestion
			return nil, errors.Errorf("Cluster for kubeconfig %q and/or context %q does not exist", i.kubeconfig.Path, i.kubeconfig.Context)
```

vincepri commented on cmd/clusterctl/client/client_test.go:
```diff
@@ -118,23 +119,23 @@ func newFakeClient(configClient config.Client) *fakeClient {
-				return nil, errors.Errorf("Repository for kubeconfig %q does not exists.", provider.ManifestLabel())
+		InjectRepositoryFactory(func(input RepositoryClientFactoryInput) (repository.Client, error) {
+			if _, ok := fake.repositories[input.provider.ManifestLabel()]; !ok {
+				return nil, errors.Errorf("Repository for kubeconfig %q does not exists.", input.provider.ManifestLabel())
```
```suggestion
				return nil, errors.Errorf("Repository for kubeconfig %q does not exist", input.provider.ManifestLabel())
```

vincepri commented on cmd/clusterctl/client/yamlprocessor/processor.go:
```diff
@@ -0,0 +1,35 @@
+
+	// ArtifactName returns the name of the template artifacts that need to be
+	// retrieved from the source.
+	ArtifactName(version, flavor string) string
```
+1, also maybe prefix with `Get` to be consistent?

wfernandes replied:
Yup...I see what you are saying. 
Since this method is technically returning the name of the "cluster" template to be used it would make sense to not be generic. I think I was using artifact because I was thinking about tar.gz blobs and didn't think of them as the actual templates.

I'll update this. Thanks!

wfernandes commented on cmd/clusterctl/client/yamlprocessor/template_processor_simple_test.go:
```diff
@@ -0,0 +1,158 @@
+		g.Expect(p.ArtifactName("", "")).To(Equal("cluster-template.yaml"))
+	})
+
+	t.Run("returns the variables from the yaml", func(t *testing.T) {
```
Yup...makes sense. Since there is no shared setup/teardown code as per the go testing patterns it makes sense to split them into separate Test funcs. For others context: https://blog.golang.org/subtests

I'll update this.

vincepri commented:
/milestone v0.3.7

wfernandes commented:
@fabriziopandini @vincepri I've addressed all of the comments. Thanks for the great feedback!

Whenever y'all get the chance please review the **last commit** that put in. 
https://github.com/kubernetes-sigs/cluster-api/pull/3115/commits/455c4be8f738120c9d86acca69778f47a260bdd2

This one changes the function prototype of the `Process` method to not depend on `config.VariablesClient`. 
See this feedback comment thread: https://github.com/kubernetes-sigs/cluster-api/pull/3115#discussion_r444095577

wfernandes commented:
/hold cancel

fabriziopandini commented:
@wfernandes outstanding job!
/lgtm

ncdc commented:
/approve

vincepri commented:
/test pull-cluster-api-test

vincepri commented:
We need to wait for #3253 to merge before tests here can pass

ncdc commented:
/retest
